### PR TITLE
Add drag and drop in EntryEditor 

### PR DESCRIPTION
**Pull request by wy8881**
fixes https://github.com/koppor/jabref/issues/569

### Mandatory checks
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [ ] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by wy8881** [comment]
Hi,
this is for fixes https://github.com/koppor/jabref/issues/569
I add a function so that a paper can be added into a group like:
1. 

2.
image
3. Can add this paper into a group when it already in another group
image

 However, this drag or drop will happen when dragging the group to the EntryEditor panel instead of dragging it to the exact group field inside General Tab. I try to add setOnDragDropped() function to a table, but it seems I can't do it. It'll be helpful if someone can give me some advice.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> I try to add setOnDragDropped() function to a tab, but it seems I can't do it.

First of all, I don't know if there is a correct way to do this X)
I would start looking at what happens if you use the components of a tab, e.g., `tab.getContent()` will give you a node.

Why a tab and not a field? If you only set it on the tab, I would be surprised if it works if you drag it to a field.

In order to deal with drag and drop to a field, I'd start looking at

https://github.com/JabRef/jabref/blob/424ff1c7cac2db2cc8d4321253bf6614cd4984b3/src/main/java/org/jabref/gui/entryeditor/EntryEditor.java#L262-L265

And if you follow that trail to its superclass `FieldsEditorTab` and

https://github.com/JabRef/jabref/blob/424ff1c7cac2db2cc8d4321253bf6614cd4984b3/src/main/java/org/jabref/gui/entryeditor/FieldsEditorTab.java#L112-L113

inside `getForField`, you'll find some fields that have special properties.I might be wrong, but I suspect you'll have to create a new field that supports dropped events.

File: src/main/java/org/jabref/gui/entryeditor/EntryEditor.java [hunk 1]
```diff
@@ -143,6 +148,18 @@ public EntryEditor(LibraryTab libraryTab) {
 
         this.setOnDragDropped(event -> {
             BibEntry entry = this.getEntry();
+            if (event.getDragboard().hasContent(DragAndDropDataFormats.GROUP)) {
+               List<String> group = (List<String>) event.getDragboard().getContent(DragAndDropDataFormats.GROUP);
+               String changedGroup = null;
+                if (entry.getField(StandardField.GROUPS).isPresent()) {
+                    changedGroup = entry.getField(StandardField.GROUPS).get() + "," + group.get(0);
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/entryeditor/EntryEditor.java:155]: 1. This is more readable using `map` in optional, i.e., `changedGroup = entry.getField(...).map().orElse();
2. JabRef allows you to change the separator, see if you can use `preferences.getBibEntryPreferences().getKeywordSeparator()` or something similar instead of `", "

Comment (wy8881) [reply src/main/java/org/jabref/gui/entryeditor/EntryEditor.java]: Thank you, I will fix it. Now I have already got everything I want. if the checks are all right, I will post an official PR.

**Comment by wy8881** [comment]
> 

Hi, I didn't find this GroupEditor. Do you mean to create a new class implementing SimpleEditor?

**Comment by calixtus** [comment]
grafik

**Comment by wy8881** [comment]
> grafik

I see. I wonder where I should put this file. It is in gui-entryeditor or  gui-fieldeditors?

**Comment by wy8881** [comment]
Hi,
I have created a GroupEditor in gui-fieldeditors. I set the drag events in bindToEntry() function since I need the entry object. I am not sure if it is appropriate. If anyone has advice please let me know. Thanks

**Comment by calixtus** [comment]
Yes I think that should work in bindToEntry. Although this method is always called if any entry in the main table is selected, there are probably no memory leaks produced, as far as I can see. Looks good so far. Let's see what a second maintainer thinks.
Thanks!

**Review by calixtus — COMMENTED** [review]
Undo unnecessary whitespace changes

**Comment by calixtus** [comment]
Just fixed some unnecessary whitespace changes. Something was not right in my ui.

**Comment by Siedlerchr** [comment]
I fear the listeners are registered everytime. I would try to register the listeners in the constructor and add a field for the entry
e.g. and in bind to entry you assign this then:

this.entry = entry;

then you have a chance to access the entry in the listeners

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> then you have a chance to access the entry in the listeners

`LinkedFilesEditor.java` follows this approach, except that it doesn't use the `SimpleEditor`. It should be possible to adapt the approach to use `SimpleEditor` as base.

**Comment by Siedlerchr** [comment]
> > then you have a chance to access the entry in the listeners
> 
> `LinkedFilesEditor.java` follows this approach, except that it doesn't use the `SimpleEditor`. It should be possible to adapt the approach to use `SimpleEditor` as base.

That is even more elegant! Yeah that should be working fine.

**Comment by wy8881** [comment]
Hi, I have removed the listener to the constructor, but I still have few questions:
1. In this code I use if to decide whether the entry has been bound instead of using .map() because that boolean variable cannot be changed inside the lambda. I'm not sure if using .map() and atomicboolean is a better idea.
https://github.com/JabRef/jabref/blob/18b204ec8a3b2e98785d4f9207501354f960a8f6/src/main/java/org/jabref/gui/fieldeditors/GroupEditor.java#L38-L44

3. After checking the LinkedFilesEditor.java, I find out that it uses ObservableOptionalValue<BibEntry> bibEntry
https://github.com/JabRef/jabref/blob/18b204ec8a3b2e98785d4f9207501354f960a8f6/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java#L70
and bind the entry in
https://github.com/JabRef/jabref/blob/18b204ec8a3b2e98785d4f9207501354f960a8f6/src/main/java/org/jabref/gui/fieldeditors/LinkedFilesEditor.java#L238
I write a similar code in GroupEditor, but it fails  noDiscardingAChangeLeadsToNewerBackupBeReported() test. The strange thing is that this can be passed in Intellij. Then I changed the type of bibEntry to Optional<BibEntry> and it can pass the test. I don't know what is wrong with the another one.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> 1. In this code I use if to decide whether the entry has been bound instead of using .map() because that boolean variable cannot be changed inside the lambda. I'm not sure if using .map() and atomicboolean is a better idea.

I am sorry, but I don't understand this explanation. Could you write a short snippet showcasing when it cannot be changed? I don't know what circumstances you cannot either use `!` or `Predicate.not`.

> 2. ... I don't know what is wrong with the another one.

This is really weird, and shouldn't happen.
Looking at the code again, perhaps `Optional<BibEntry>` is the better option. In this setting, there is no need to use JavaFX bindings at all.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> 1. In this code I use if to decide whether the entry has been bound instead of using .map() because that boolean variable cannot be changed inside the lambda. I'm not sure if using .map() and atomicboolean is a better idea.

You'll have difficulties dealing with local variables inside of a lambda. If you are willing to try you can rewrite almost the whole dragDropped function and end it with `.ifPresentOrElse` using `event.setDropCompleted(true/false)`.

It would be an interesting exercise, but I don't know if it would make it unreadable.

**Comment by wy8881** [comment]
> It would be an interesting exercise, but it might make it unreadable.
> I tried to rewrite the setOnDragDropped and it seems ugly, so I think it's better to stick with 'if' version
> The outer `if` would most likely have to be changed to an `if-else`
I can change that but I don't understand the point. I think if the condition fails and the boolean success will stay false. I'm not sure why we still need the else clause.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> I tried to rewrite the setOnDragDropped and it seems ugly, so I think it's better to stick with 'if' version

Agreed, thank you for humoring me, and sorry :sweat_smile: 

> I can change that but I don't understand the point. I think if the condition fails and the boolean success will stay false. I'm not sure why we still need the else clause.

Logically nothing should change, true. But from my (opinionated :grin:) point of view, keeping each `if` as short/small as possible often creates less code complexity, even if the "logic condition" complexity is the same in this case.

I just re-arranged your latest code and arrived at this, and just to be clear, I think the first version you made with an `if` is easier to read, and I'd suggest you revert to that version.

Here I can mentally deal with the `if (event.getDragboard().hasContent(DragAndDropDataFormats.GROUP))` before I start checking on what the `bibEntry` code does. Imo it makes it easier to see if there is a missing `event.setDropCompleted(false)` (line 43 :wink:)

```java
          if (event.getDragboard().hasContent(DragAndDropDataFormats.GROUP)) {
                List<String> group = (List<String>) event.getDragboard().getContent(DragAndDropDataFormats.GROUP);
                bibEntry.ifPresentOrElse(entry -> {
                    final String changedGroup = entry.getField(StandardField.GROUPS)
                                                     .map(setGroup -> setGroup + (preferences.getBibEntryPreferences().getKeywordSeparator()) + (group.get(0))).orElse(group.get(0));
                    entry.setField(StandardField.GROUPS, changedGroup);
                    event.setDropCompleted(true);
                }, () -> event.setDropCompleted(false));
            } else {
                event.setDropCompleted(false);
            }
            event.consume();
        });
```

File: src/main/java/org/jabref/gui/fieldeditors/GroupEditor.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+            }
+            event.consume();
+        });
+
+        this.setOnDragDropped(event -> {
+            bibEntry.ifPresentOrElse((entry) -> {
+                if (event.getDragboard().hasContent(DragAndDropDataFormats.GROUP)) {
+                    List<String> group = (List<String>) event.getDragboard().getContent(DragAndDropDataFormats.GROUP);
+                    String changedGroup = entry.getField(StandardField.GROUPS)
+                                               .map(setGroup -> setGroup + (preferences.getBibEntryPreferences().getKeywordSeparator()) + (group.get(0)))
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/fieldeditors/GroupEditor.java:39]: Even after you revert these changes, I'd suggest changing some names to clarify,

- `List<String> group` -> `List<String> draggedGroups` (or `draggedGroup` if you add `.get(0)` to event.getDragboard().getContent(DragAndDropDataFormats.GROUP))
- `setGroup` -> `oldGroups`

Or if you believe something else is better

Comment (wy8881) [reply src/main/java/org/jabref/gui/fieldeditors/GroupEditor.java]: > Here I can mentally deal with the `if (event.getDragboard().hasContent(DragAndDropDataFormats.GROUP))` before I start checking on what the `bibEntry` code does. Imo it makes it easier to see if there is a missing `event.setDropCompleted(false)` (line 43 😉)
> 
> ```java
>           if (event.getDragboard().hasContent(DragAndDropDataFormats.GROUP)) {
>                 final String group = (List<String>) event.getDragboard().getContent(DragAndDropDataFormats.GROUP).get(0);
>                 bibEntry.ifPresentOrElse(entry -> {
>                     final String changedGroup = entry.getField(StandardField.GROUPS)
>                                                      .map(setGroup -> setGroup + (preferences.getBibEntryPreferences().getKeywordSeparator()) + (group)).orElse(group);
>                     entry.setField(StandardField.GROUPS, changedGroup);
>                     event.setDropCompleted(true);
>                 }, () -> event.setDropCompleted(false));
>             } else {
>                 event.setDropCompleted(false);
>             }
>             event.consume();
>         });
> ```

Thank you for your explanation and it is inspiring. I think if-else is better with longer code, and people don't have to search the whole way down to find where 'success' is changed. I will keep it in mind when dealing with similar issues in the future😁. I have changed the name of the variables in my newest commit.

**Review by calixtus — DISMISSED** [review]

File: src/main/java/org/jabref/gui/fieldeditors/GroupEditor.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        });
+
+        this.setOnDragDropped(event -> {
+            boolean success = false;
+            if (event.getDragboard().hasContent(DragAndDropDataFormats.GROUP)) {
+                List<String> draggedGroups = (List<String>) event.getDragboard().getContent(DragAndDropDataFormats.GROUP);
+                if (bibEntry.isPresent()) {
+                    String newGroup = bibEntry.map(entry -> entry.getField(StandardField.GROUPS)
+                                                                     .map(oldGroups -> oldGroups + (preferences.getBibEntryPreferences().getKeywordSeparator()) + (draggedGroups.get(0)))
+                                                                     .orElse(draggedGroups.get(0)))
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/gui/fieldeditors/GroupEditor.java:41]: Add a check that `draggedGroups.get(0)` isn't an empty string. If you drag the "All Entries" group in JabRef it will keep appending separators.

**Review by k3KAW8Pnf7mkmdSMPHz27 — CHANGES_REQUESTED** [review]
Codewise, it looks good to me! Two very minor things and it is ready to go,

1. Add a short note to the CHANGELOG
2. Address the empty string case for groups (dragging all entries, see note in code)

**Review by calixtus — APPROVED** [review]

**Comment by wy8881** [comment]
Thank you for the help
